ui/pages/confirmations/hooks/alerts/useConfirmationOriginAlerts.ts

```diff
@@ -0,0 +1,49 @@
+import { useMemo } from 'react';
+import { useSelector } from 'react-redux';
+import { TransactionMeta } from '@metamask/transaction-controller';
+
+import { Alert } from '../../../../ducks/confirm-alerts/confirm-alerts';
+import { Severity } from '../../../../helpers/constants/design-system';
+import { useI18nContext } from '../../../../hooks/useI18nContext';
+import { Confirmation, SignatureRequestType } from '../../types/confirm';
+import { currentConfirmationSelector } from '../../selectors';
+import { isValidASCIIURL, toPunycodeURL } from '../../utils/confirm';
+import { isSignatureTransactionType } from '../../utils';
+
+const useConfirmationOriginAlerts = (): Alert[] => {
+  const t = useI18nContext();
+
+  const currentConfirmation: Confirmation | undefined = useSelector(
+    currentConfirmationSelector,
+  );
+
+  return useMemo<Alert[]>((): Alert[] => {
+    if (!currentConfirmation) {
+      return [];
+    }
+
+    const origin = isSignatureTransactionType(currentConfirmation)
+      ? (currentConfirmation as SignatureRequestType).msgParams?.origin
+      : (currentConfirmation as TransactionMeta).origin;
+
+    if (origin && isValidASCIIURL(origin)) {
+      return [];
+    }
+
+    return [
+      {
+        key: 'originSpecialCharacterWarning',
+        reason: t('addressMismatch'),
+        field: 'originSpecialCharacterWarning',
+        severity: Severity.Warning,
+        message: t('addressMismatchWarning'),
+        alertDetails: [
+          t('addressMismatchOriginal', [origin]),
+          t('addressMismatchPunycode', [origin ? toPunycodeURL(origin) : '']),
+        ],
+      },
+    ];
+  }, [currentConfirmation, t]);
```

Comment:
Rather than returning a fresh object every time the confirmation changes at all, could we instead use the most stable dependencies possible such as `isInvalidAscii`, `field` and `t`?

It also simplifies the code since we can move the core logic out of the `useMemo`.